File: Python/src/Cracking the Coding Interview Problems/One_One.py
"""1.1 Implement an algorithm to determine if a string has all unique characters. What
if you cannot use additional data structures?"""


def all_unique_characters(string):
    hash_tab = [0 for j in range(0, 500)]
    for i in range(0, string.__len__() - 1):
        if hash_tab[ord(string[i])] == 0:
            hash_tab[ord(string[i])] = 1
        else:
            return False
    return True


def main():
    i = raw_input("Enter a string to be tested: ")
    print all_unique_characters(i)
    return


main()

File: Java/src/Cracking The Coding Interview/One_One.java
package Ch1_InterviewQuestions;

import java.util.Scanner;

/*
    1.1 Implement an algorithm to determine if a string has all unique characters. What
	if you cannot use additional data structures?
 */

public class One_One 
{
	public static void main(String[] args)
	{
		Scanner in = new Scanner(System.in);
		String input = in.nextLine();
		System.out.println(hasUniqueCharacters(input));
	}
	
	private static boolean hasUniqueCharacters(String input)
	{
		char a, b;
		for(int i = 0; i < input.length(); i++)
		{
			for(int j = 0; j < input.length(); j++)
			{
				if(j != i)
				{
					a = input.charAt(i);
					b = input.charAt(j);
					if(a == b)
					{
						return false;
					}
				}
				else
				{
					continue;
				}
			}
		}
		return true;
	}
}

//VERY inefficient, as the worst case is O(n^2). Needs to be an easier way just am not sure right now
